File: packages/meal-plan-engine/src/filters.ts
import type { Meal, Allergen, DietaryTag, MealCategory } from "@caloriehero/shared-types";

// ---------------------------------------------------------------------------
// Allergen filtering
// ---------------------------------------------------------------------------

/**
 * Removes meals that contain any of the listed allergens.
 * Strict: no false negatives — if a meal lists the allergen it is excluded.
 */
export function filterByAllergens(meals: Meal[], allergies: Allergen[]): Meal[] {
  if (allergies.length === 0) return meals;
  const allergenSet = new Set<Allergen>(allergies);
  return meals.filter(
    (meal) => !meal.allergens.some((a) => allergenSet.has(a))
  );
}

// ---------------------------------------------------------------------------
// Dietary tag filtering
// ---------------------------------------------------------------------------

/**
 * Keeps only meals that have ALL listed dietary tags (AND logic).
 */
export function filterByDietaryTags(
  meals: Meal[],
  preferences: DietaryTag[]
): Meal[] {
  if (preferences.length === 0) return meals;
  return meals.filter((meal) =>
    preferences.every((tag) => meal.dietaryTags.includes(tag))
  );
}

// ---------------------------------------------------------------------------
// Category filtering
// ---------------------------------------------------------------------------

/**
 * Keeps only meals belonging to the given category.
 */
export function filterByCategory(meals: Meal[], category: MealCategory): Meal[] {
  return meals.filter((meal) => meal.category === category);
}

// ---------------------------------------------------------------------------
// Combined filter pipeline
// ---------------------------------------------------------------------------

export interface FilterOptions {
  allergies?: Allergen[];
  dietaryPreferences?: DietaryTag[];
  category?: MealCategory;
}

/**
 * Runs all applicable filters in sequence: allergens → dietary tags → category.
 */
export function filterMeals(meals: Meal[], opts: FilterOptions): Meal[] {
  let result = meals;

  if (opts.allergies && opts.allergies.length > 0) {
    result = filterByAllergens(result, opts.allergies);
  }

  if (opts.dietaryPreferences && opts.dietaryPreferences.length > 0) {
    result = filterByDietaryTags(result, opts.dietaryPreferences);
  }

  if (opts.category !== undefined) {
    result = filterByCategory(result, opts.category);
  }

  return result;
}

File: backend/app/engine/filters.py
"""Meal filtering functions."""

from app.engine.types import Meal


def filter_by_allergens(meals: list[Meal], allergies: list[str]) -> list[Meal]:
    """Remove meals containing any of the listed allergens."""
    if not allergies:
        return meals
    allergen_set = set(allergies)
    return [m for m in meals if not any(a in allergen_set for a in m.allergens)]


def filter_by_dietary_tags(meals: list[Meal], preferences: list[str]) -> list[Meal]:
    """Keep only meals that have ALL listed dietary tags (AND logic)."""
    if not preferences:
        return meals
    return [m for m in meals if all(tag in m.dietary_tags for tag in preferences)]


def filter_by_category(meals: list[Meal], category: str) -> list[Meal]:
    """Keep only meals belonging to the given category."""
    return [m for m in meals if m.category == category]


def filter_meals(
    meals: list[Meal],
    allergies: list[str] | None = None,
    dietary_preferences: list[str] | None = None,
    category: str | None = None,
) -> list[Meal]:
    """Run all applicable filters in sequence: allergens -> dietary tags -> category."""
    result = meals

    if allergies:
        result = filter_by_allergens(result, allergies)

    if dietary_preferences:
        result = filter_by_dietary_tags(result, dietary_preferences)

    if category is not None:
        result = filter_by_category(result, category)

    return result
